### PR TITLE
Covered my owm HashSet with jUnit tests

### PR DESCRIPTION
**Pull request by dmurasov**
Added jUnit tests for HashSet

### CODE REVIEW

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+    @BeforeEach
+    void setUpStudentSet(){
+        set = new StudentSet();
+    }
+
+    @Test
+    void addElementShouldReturnTrue(){
+        Student st = new Student("John Doe", LocalDate.of(2000, 5, 12), "");
+        assertTrue(set.add(st));
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: assert colection contains student and size = 1

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Test
+    void addDuplicatedNullToSetShouldAllowASingleNullElement(){
+        set.add(null);
+        set.add(null);
+        assertThat(set).containsNull();
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: assert size

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Test
+    void addDublicatedObjectToSetShouldNotBeAdded(){
+        Student st = new Student("John Doe", LocalDate.of(2000, 5, 12), "");
+        set.add(st);
+        assertAll(
+                () -> assertFalse(set.add(st)),
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: move please all calls to the hashSet's methods outside of asserts ( from all the methods)

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    @Test
+    void removeNullElementThatIsNotContainedShouldReturnFalse(){
+        assertFalse(set.remove(null));
+    }
+
+    @Test
+    void nullElementShouldBeRemovedIfItIsContained(){
+        set.add(null);
+
+        set.remove(null);
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: also assert what remove returns

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        assertFalse(set.contains(null));
+    }
+
+    @Test
+    void methodShouldReturnIteratorThatContainsNotNullElementsFromTheSet(){
+        Student st1 = new Student("John Doe", LocalDate.of(2000, 5, 12), "");
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: move stundents to be constants, and rename st1 to JOHN

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Test
+    void shouldReturnTrueIfAllElementsFromOtherCollectionIsContainedInSetInAnyPosition(){
+        Student st1 = new Student("John Doe", LocalDate.of(2000, 5, 12), "");
+        Student st2 = new Student("David", LocalDate.of(2001, 6, 20), "");
+
+        addElementsToSet(set, st1, st2, null);
+
+        List<Student> list = Arrays.asList(st1, null);
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: make all variables final if possible

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 1 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Fixed test

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 2 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Fixed test

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 3 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Fixed test

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 4 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Fixed test

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 5 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Fixed test

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 6 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: All variables are made final

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        set = new StudentSet(1);
+        addElementsToSet(set, st1, st2, st3);
+
+        assertThat(set).hasSize(3);
+    }
+
+
+    private void addElementsToSet(StudentSet set, final Student ... students){
+        for(Student s : students){
+            assertTrue(set.add(s));
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: this is a helper method, please remove this assert

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 6 above)
Comment (perepiolcamaxim) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: **final** List<Student> list = Arrays.asList(st1, null);

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+    private final static Student st1 = new Student("John", LocalDate.of(2000, 5, 12), "");
+
+    private final static Student st2 = new Student("David", LocalDate.of(2001, 6, 20), "");
+
+    private final static Student st3 = new Student("Dmitrii", LocalDate.of(1996, 10, 16), "");
+
+    @BeforeEach
+    void setUpStudentSet(){
+        set = new StudentSet();
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: remove this setup method, and initialise in line like this :  
private StudentSet set = new StudentSet();

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+import java.time.LocalDate;
+
+public class StudentSetTest {
+    private StudentSet set;
+
+    private final static Student st1 = new Student("John", LocalDate.of(2000, 5, 12), "");
+
+    private final static Student st2 = new Student("David", LocalDate.of(2001, 6, 20), "");
+
+    private final static Student st3 = new Student("Dmitrii", LocalDate.of(1996, 10, 16), "");
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: rename please this constants to JOHN, DAVID and DMITRII

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    void nullElementShouldBeRemovedIfItIsContained(){
+        assertTrue(set.add(null));
+        assertTrue(set.remove(null));
+        assertFalse(set.contains(null));
+    }
+
+    @Test
+    void methodShouldReturnIteratorThatContainsNotNullElementsFromTheSet(){
+        addElementsToSet(set, st1, st2);
+        Iterator it = set.iterator();
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: **final**  Iterator it = set.iterator();

File: src/test/java/com/endava/internship/collections/StudentSetTest.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        while(it.hasNext()) {
+            assertNotNull(it.next());
+        }
+    }
+
+    @Test
+    void elementsFromOtherCollectionShouldBeAddedInExactlyOrder(){
+        List<Student> list = Arrays.asList(st1, null, st2);
```
Comment (perepiolcamaxim) [inline src/test/java/com/endava/internship/collections/StudentSetTest.java]: **final** List<Student> list = Arrays.asList(st1, null, st2);

Comment (perepiolcamaxim) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: plese add final to all variables in this class

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 7 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java:214]: If I need it to be cleared or created new set after each test?

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 8 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: If I need it to be cleared or created new set after each test?

Comment (perepiolcamaxim) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: https://www.baeldung.com/junit-testinstance-annotation

Comment (perepiolcamaxim) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: check  example from section 2 Default Test Lifecycle

Comment (perepiolcamaxim) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: on every test method, a new Set object is created

**Review by perepiolcamaxim — APPROVED** [review]

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 11 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Variables added

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 10 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Modified

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 9 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Constants have been renamed

File: src/test/java/com/endava/internship/collections/StudentSetTest.java (hunk 7 above)
Comment (dmurasov) [reply src/test/java/com/endava/internship/collections/StudentSetTest.java]: Assert was removed for this method

**Pull request merged**